programs/drift_vaults/src/state/vault.rs

```diff
@@ -504,4 +738,391 @@ impl Vault {
+        let vault_shares_after = self.get_protocol_shares(vault_protocol);
+
+        if let Some(vp) = vault_protocol {
+            self.total_withdraw_requested = self
+                .total_withdraw_requested
+                .safe_sub(vp.last_protocol_withdraw_request.value)?;
+            vp.last_protocol_withdraw_request.reset(now)?;
+
+            self.emit_vault_depositor_record(
+                VaultDepositorRecordParams {
+                    ts: now,
+                    action: VaultDepositorAction::CancelWithdrawRequest,
+                    amount: 0,
+                    depositor_authority: vp.protocol,
+                    vault_equity_before: vault_equity,
+                    vault_shares_before,
+                    user_vault_shares_before,
+                    total_vault_shares_before,
+                    vault_shares_after,
+                    manager_profit_share: 0,
+                    management_fee: management_fee_payment,
+                    management_fee_shares,
+                },
+                Some(VaultDepositorRecordProtocolParams {
+                    protocol_profit_share: 0,
+                    protocol_fee: protocol_fee_payment,
+                    protocol_fee_shares,
+                }),
+            )?;
+        }
+
+        Ok(())
+    }
+
+    pub fn protocol_withdraw(
+        &mut self,
+        vault_protocol: &mut Option<RefMut<VaultProtocol>>,
+        vault_equity: u64,
+        now: i64,
+    ) -> Result<u64> {
+        if vault_protocol.is_none() {
+            validate!(
+                false,
+                ErrorCode::VaultProtocolMissing,
+                "Protocol cannot withdraw for a non-protocol vault"
+            )?;
+        }
+
+        self.last_manager_withdraw_request
+            .check_redeem_period_finished(self, now)?;
+
+        self.apply_rebase(vault_protocol, vault_equity)?;
+
+        let VaultFee {
+            management_fee_payment,
+            management_fee_shares,
+            protocol_fee_payment,
+            protocol_fee_shares,
+        } = self.apply_fee(vault_protocol, vault_equity, now)?;
+
+        let vault_shares_before: u128 = self.get_protocol_shares(vault_protocol);
+        let total_vault_shares_before = self.total_shares;
+        let user_vault_shares_before = self.user_shares;
+
+        let n_shares = match vault_protocol {
+            None => 0,
+            Some(vp) => vp.last_protocol_withdraw_request.shares,
+        };
+
+        validate!(
+            n_shares > 0,
+            ErrorCode::InvalidVaultWithdraw,
+            "Must submit withdraw request and wait the redeem_period ({} seconds)",
+            self.redeem_period
+        )?;
+
+        let amount: u64 =
+            depositor_shares_to_vault_amount(n_shares, self.total_shares, vault_equity)?;
+
+        let n_tokens = match vault_protocol {
+            None => amount,
+            Some(vp) => amount.min(vp.last_protocol_withdraw_request.value),
+        };
+
```

Comment:
unnecessary check since done at the beginning of function

Reply:
I don't understand. This pattern is done many times as the `vault_protocol` can't be unwrapped since it's behind a mutable reference and needs to be passed as the Option to multiple functions.

So there's no choice but to `match` or `Option::<T>.as_ref().map(...).unwrap_or(...)`
Either way it matches under the hood.

If you ctrl+f for `match vault_protocol` you'll see the same pattenr in many places.

Reply:
What do you mean? I need to grab the n_tokens value, and that's not done earlier in the fn?